system/adb/shell_exec_builtin.c

```diff
@@ -0,0 +1,200 @@
+/****************************************************************************
+ * system/adb/shell_pipe_uv.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <spawn.h>
+#include <fcntl.h>
+#include <errno.h>
+
+#include "adb.h"
+#include "builtin/builtin.h"
+#include "shell_pipe.h"
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+int shell_exec_builtin(FAR const char *appname, FAR char * const *argv,
+                       FAR shell_pipe_t *apipe)
+{
+  FAR const struct builtin_s *builtin;
+  posix_spawnattr_t attr;
+  posix_spawn_file_actions_t file_actions;
+  struct sched_param param;
+  pid_t pid;
+  int index;
+  int ret;
+
+  /* Verify that an application with this name exists */
+
+  index = builtin_isavail(appname);
```

Comment:
should we support to launch program on file system?

Reply:
Indeed this can be simplified. This function is the same as the one in ./builtin/exec_builtin.c but with more "posix_spawn_file_actions_addopen()" to setup correctly stdin and stdout streams.